python-frontend/src/main/java/org/sonar/python/semantic/v2/ProjectLevelTypeTable.java

```diff
@@ -42,48 +42,76 @@
 
   public ProjectLevelTypeTable(ProjectLevelSymbolTable projectLevelSymbolTable) {
     this.projectLevelSymbolTable = projectLevelSymbolTable;
-    this.modules = new HashMap<>();
-    Map<String, Symbol> builtinSymbols = TypeShed.builtinSymbols();
-    this.builtinTypes = new HashMap<>();
-    builtinSymbols.forEach((key, value) -> builtinTypes.put(key, convertToType(value)));
-    modules.put("builtins", new ModuleType("builtins", builtinTypes));
+    this.rootModule = new ModuleType(null);
+    initBuiltinModule();
   }
 
-  public ModuleType getModule(String moduleName) {
-    if (modules.containsKey(moduleName)) {
-      return modules.get(moduleName);
+  public ModuleType getModule(String... moduleName) {
+    return getModule(List.of(moduleName));
+  }
+
+  public ModuleType getModule(List<String> moduleNameParts) {
+    var parent = rootModule;
+    for (int i = 0; i < moduleNameParts.size(); i++) {
+      var existing = parent.resolveMember(moduleNameParts.get(i));
+
+      if (existing instanceof ModuleType existingModule) {
+        parent = existingModule;
+        continue;
+      }
+
+      var moduleFqn = IntStream.rangeClosed(0, i)
+        .mapToObj(moduleNameParts::get)
+        .collect(Collectors.joining("."));
+
+      parent = initModule(moduleNameParts.get(i), moduleFqn, parent);
     }
-    Set<Symbol> symbolsFromModule = projectLevelSymbolTable.getSymbolsFromModule(moduleName);
-    Map<String, PythonType> children;
-    if (symbolsFromModule != null) {
-      children = symbolsFromModule.stream()
-        .map(this::convertToType)
-        .collect(Collectors.toMap(PythonType::name, Function.identity(), (a, b) -> PythonType.UNKNOWN));
-    } else {
-      // FIXME: what to do here?
-      children = Collections.emptyMap();
+    return parent;
+  }
+
+  private ModuleType initModule(String moduleName, String moduleFqn, ModuleType parent) {
+    var projectModuleSymbols = projectLevelSymbolTable.getSymbolsFromModule(moduleFqn);
+    if (projectModuleSymbols != null) {
+      return createModuleFromSymbols(moduleName, parent, projectModuleSymbols);
     }
 
-    ModuleType moduleType = new ModuleType(moduleName, children);
-    modules.put(moduleName, moduleType);
-    return moduleType;
+    var typeShedModuleSymbols = TypeShed.symbolsForModule(moduleFqn);
+    if (typeShedModuleSymbols != null) {
+      return createModuleFromSymbols(moduleName, parent, typeShedModuleSymbols);
+    }
+
+    var emptyModule = new ModuleType(moduleName, parent);
+    parent.members().put(moduleName, emptyModule);
+    return emptyModule;
   }
 
-  private PythonType convertToType(Symbol symbol) {
-    return switch (symbol.kind()) {
-      case CLASS -> converToClassType((ClassSymbol) symbol);
-      case FUNCTION -> convertToFunctionType((FunctionSymbol) symbol);
-      case AMBIGUOUS -> convertToUnionType((AmbiguousSymbol) symbol);
-      case OTHER -> converToObjectType(symbol);
-    };
+  private void initBuiltinModule() {
```

Comment:
I do have some doubts about how we eventually merge stub symbols from project symbols in the project table.

We used to have a somewhat clear order of priority in how we resolve symbols (project symbols > custom typeshed > stdlib typeshed) that I feel we may lose. I'd be tempted to have a project `ModuleType` and a stub `ModuleType` instead, wdyt?

Still, I think it's a good idea to centralize those symbols/types in the project table (as opposed to the previous implementation where we had project symbols in the project table and Typeshed symbols in the Typeshed class, sequentially looking in each location at the import resolution).

Reply:
As discussed the priority should be handled not as different storages but on type population stage. 
Maybe it will make sense to introduce a type source field for PythonType but I suggest adding it only once it is really used

Reply:
That makes sense. I must say I'm still partial to having clearly separated storages, but at the moment I can't really express a compelling "why" (it's just a gut feeling).

Let's keep things as they are now and keep this option as a possibility if we face limitations.

As I'm typing this, one thing we may want to double-check is how PR analysis caching works with project symbols vs. stub symbols. I tend to think we don't want to cache stub symbols (due to the network cost and them being embedded in the analyzer anyway) however I know we had problems in the past with this (because stub symbols depend on which files are actually parsed). We don't need to act on this as of now, this is just a consideration I wanted to share.